text/0000-engines-component-dependencies.md

```diff
@@ -0,0 +1,97 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+This RFC proposes to allow the declaration of components as dependencies of engines.
+
+# Motivation
+
+Currently, the only way to allow engines to share components with their consuming application/engine is to extract the component and package it as a standalone add-on. The add-on can then be installed by any application/engine that requires it.
+
+While this approach is a perfectly viable option in some circumstances, the viability begins to breakdown in distributed applications where the developer may not have control of some or all of the engines being consumed.
+
+In such circumstances it becomes difficult to ensure consistency throughout the application as each engine could theoretically be using a different version of the same component, potentially resulting in different areas of the application having different UI/UX for the same component/functionality.
```

Comment:
Our use-case for component dependencies is a CMS, the main problem that we are trying to solve, is being able to control as much of admin panels UI/UX as possible.

We want to make sure that users are provided with a familiar UX/UI where ever they are in the admin panel, be that in an officially maintained section, or within a section provided by a plugin developed and maintained by a third party.

We would also like to have the flexibility to make changes to the components we provide, and have those changes flow through the entire admin panel, even into sections that are maintained by a third party.

A simple example could be if we decided to change the HTML editor that was being used from something like Froala to Redactor.

Using component dependencies, we could update the admin panels `html-editor` component, and have that change flow through to the entire admin panel, including through to engines that had been developed by third parties.

Using a stand-alone add-on on the other hand, we would need to develop and maintain an ember add-on that plugin authors would be required to install and keep up-to-date in their engines. When we made any changes, e.g. to the `html-editor` component above, we would not only need to hope that the engine author updated their engine to use the latest version of our add-on, but then also hope that the user of the CMS updated the third party plugin to the latest version, allowing the changes we'd made to flow into the application.

Our concern here, is that users could very quickly end up with an admin panel that has a different UI/UX depending where they are in the admin panel. Using the example above, they could end up in a situation where they are using Froala as an HTML editor in one area of the admin panel, and Redactor in another section. This could lead to confusion due to the different look/feel/layout of the HTML editors.